vpr/src/place/timing_place.h

```diff
@@ -55,40 +97,148 @@ class PlacerCriticalities {
- *      [0..cluster_ctx.clb_nlist.nets().size()-1][1..num_pins-1]
+ * This class mirrors PlacerCriticalities by both its methods and its members. The only
+ * difference is that this class deals with RAW setup slacks returned by SetupTimingInfo
+ * rather than criticalities. See the documentation on PlacerCriticalities for more.
+ *
+ * RAW setup slacks are unlike criticalities. Their values are not confined between
+ * 0 and 1. Their values can be either positive or negative.
+ *
+ * This class also provides iterating over the clustered netlist connections/pins that
+ * have modified setup slacks by the last call to update_setup_slacks(). However, this
+ * utility is mainly used for incrementally committing the setup slack values into the
+ * structure `connection_setup_slack` used by many placer routines.
+ */
+class PlacerSetupSlacks {
+  public: //Types
+    typedef vtr::vec_id_set<ClusterPinId>::iterator pin_iterator;
+    typedef vtr::vec_id_set<ClusterNetId>::iterator net_iterator;
+
+    typedef vtr::Range<pin_iterator> pin_range;
+    typedef vtr::Range<net_iterator> net_range;
+
+  public: //Lifetime
+    PlacerSetupSlacks(const ClusteredNetlist& clb_nlist, const ClusteredPinAtomPinsLookup& netlist_pin_lookup);
+    PlacerSetupSlacks(const PlacerSetupSlacks& clb_nlist) = delete;
+    PlacerSetupSlacks& operator=(const PlacerSetupSlacks& clb_nlist) = delete;
+
+  public: //Accessors
+    ///@brief Returns the setup slack of the specified connection.
+    float setup_slack(ClusterNetId net, int ipin) const { return timing_place_setup_slacks_[net][ipin]; }
+
+    /**
+     * @brief Returns the range of clustered netlist pins (i.e. ClusterPinIds)
+     *        which were modified by the last call to PlacerSetupSlacks::update_setup_slacks().
+     */
+    pin_range pins_with_modified_setup_slack() const;
+
+  public: //Modifiers
+    ///@brief Updates setup slacks based on the atom netlist setup slacks provided by timing_info.
+    void update_setup_slacks(const SetupTimingInfo* timing_info);
+
+    ///@brief Override the setup slack of a particular connection.
+    void set_setup_slack(ClusterNetId net, int ipin, float val);
+
+    ///@brief Set `update_enabled` to true.
+    void enable_update() { update_enabled = true; }
+
+    ///@brief Set `update_enabled` to true.
+    void disable_update() { update_enabled = false; }
+
+  private: //Data
+    const ClusteredNetlist& clb_nlist_;
+    const ClusteredPinAtomPinsLookup& pin_lookup_;
+
+    /**
+     * @brief The matrix that stores raw setup slack values for each connection.
+     *
+     * Index range: [0..cluster_ctx.clb_nlist.nets().size()-1][1..num_pins-1]
+     */
+    ClbNetPinsMatrix<float> timing_place_setup_slacks_;
+
+    ///@brief Set of pins with raw setup slacks modified by last call to update_setup_slacks()
+    vtr::vec_id_set<ClusterPinId> cluster_pins_with_modified_setup_slack_;
+
+    ///@brief Updates setup slacks: incremental V.S. from scratch.
```

Comment:
Please carry over any commenting enhancements for the placer_criticalities based on that discussion to here as well.